src/clustering_quality.jl

```diff
@@ -0,0 +1,328 @@
+    )
+
+    n, k = size(X, 2), size(centers,2)
+    w_idx1, w_idx2 = axes(weights)
+
+    global_center = vec(mean(X, dims=2))
+    center_distances = colwise(distance, centers, global_center)
+    outer_intertia = sum(
+        weights[i,j₁]^fuzziness * center_distances[j₂] for i in w_idx1, (j₁,j₂) in zip(w_idx2, 1:k)
+    )
+
+    inner_intertia = _inner_inertia(X, centers, weights, fuzziness, distance)
+
+    return (outer_intertia / (k - 1)) / (inner_intertia / (n - k))
+end
+
+# Davies-Bouldin idex 
+
+function _cluquality_davies_bouldin(
+        X::AbstractMatrix{<:Real},
+        centers::AbstractMatrix{<:Real},
+        assignments::AbstractVector{<:Integer},
+        distance::SemiMetric=SqEuclidean()
+    )
+
+    k = size(centers, 2)
+    c_idx = axes(centers,2)
+
+    cluster_samples = _gather_samples(assignments, k)
+
+    cluster_diameters = [mean(colwise(distance,view(X, :, sample), centers[:,j])) for (j, sample) in zip(c_idx, cluster_samples) ]
+    center_distances = pairwise(distance,centers)
+
+    DB = mean(
+        maximum( (cluster_diameters[j₁] + cluster_diameters[j₂]) / center_distances[j₁,j₂] for j₂ in c_idx if j₂ ≠ j₁)
+            for j₁ in c_idx
+    )
+    return  DB
+end
+
+
+# Xie-Beni index
+
+function _cluquality_xie_beni(
+        X::AbstractMatrix{<:Real},
+        centers::AbstractMatrix{<:Real},
+        assignments::AbstractVector{<:Integer},
+        distance::SemiMetric=SqEuclidean()
+    )
+
+    n, k = size(X, 2), size(centers,2)
+
+    cluster_samples = _gather_samples(assignments, k)
+    inner_intertia  = _inner_inertia(X, centers, cluster_samples, distance)
+
+    center_distances = pairwise(distance,centers)
+    min_center_distance = minimum(center_distances[j₁,j₂] for j₁ in 1:k for j₂ in j₁+1:k)
+    
+    return inner_intertia / (n * min_center_distance)
+end
+
+function _cluquality_xie_beni(
+        X::AbstractMatrix{<:Real},
+        centers::AbstractMatrix{<:Real},
+        weights::AbstractMatrix{<:Real},
+        fuzziness::Real,
+        distance::SemiMetric=SqEuclidean()
+    )
+
+    n, k = size(X, 2), size(centers, 2)
+
+    inner_intertia = _inner_inertia(X, centers, weights, fuzziness, distance)
+
+    center_distances = pairwise(distance, eachcol(centers))
+    min_center_distance = minimum(center_distances[j₁,j₂] for j₁ in 1:k for j₂ in j₁+1:k)
+
+    return inner_intertia / (n * min_center_distance)
+end
+
+
+# Dunn index
+
+function _cluquality_dunn(assignments::AbstractVector{<:Integer}, dist::AbstractMatrix{<:Real})
+
+    k = maximum(assignments)
+
+    cluster_samples = _gather_samples(assignments, k)
+
+    min_outer_distance = minimum(
+        minimum(view(dist, cluster_samples[j₁], cluster_samples[j₂]), init = typemax(eltype(dist)))
+            for j₁ in 1:k for j₂ in j₁+1:k
+    )
+
+    max_inner_distance = maximum(
+        maximum(dist[i₁,i₂] for i₁ in sample, i₂ in sample, init = typemin(eltype(dist)))
+            for sample in cluster_samples
+    )
```

Comment:
Using `assignments` directly looks faster and does not allocate:
```suggestion
    max_inner_distance = typemin(eltype(dist))
    min_outer_distance = typemax(eltype(dist))
    for ((i, clu_i), dist_i) in zip(enumerate(assignments), eachcol(dist))
        for j in (i + 1):k
            dist = dist_i[j]
            if assignments[j] == clu_i
                if max_inner_distance > dist
                    max_inner_distance = dist
                end
            else
                if min_outer_distance < dist
                    min_outer_distance = dist
                end
            end
        end
    end
```

Reply:
Ah, there's a bug in my code: the j loop should be `for j in (i + 1):n`, where `n = length(assignments)`

Reply:
we just iterate over all pairs of points and update max inner distance if the pair is in the same cluster or min outer distance otherwise